src/sd.cpp

```diff
@@ -1091,36 +1093,146 @@ inline static std::vector<std::pair<std::string, float>> parse_prompt_attention(
     return res;
 }
 
-inline static std::vector<std::string> split(std::string str)
+struct pair_hash
 {
-    std::string::size_type pos;
-    std::vector<std::string> result;
-    str += " ";
-    int size = str.size();
+    std::size_t operator() (const std::pair<std::string, std::string> &pair) const
+    {
+        return std::hash<std::string>{}(pair.first) ^ std::hash<std::string>{}(pair.second);
+    }
+};
+
+inline static std::unordered_set<std::pair<std::string, std::string>, pair_hash> get_pairs(std::vector<std::string> word)
+{
+    std::unordered_set<std::pair<std::string, std::string>, pair_hash> pairs;
+    std::string prev_char = word[0];
+
+    for (size_t i = 1; i < word.size(); ++i) {
+        pairs.insert({prev_char, word[i]});
+        prev_char = word[i];
+    }
 
-    for (int i = 0; i < size; i++)
+    return pairs;
+}
+
+inline static std::vector<std::string> bpe(std::string str, std::unordered_map<std::pair<std::string, std::string>, int, pair_hash>& tokenizer_bperankings)
+{
+    std::vector<std::string> word;
+    for (size_t i = 0; i < str.size() - 1; ++i)
+        word.push_back(std::string(1, str[i]));
+    word.push_back(std::string(1, str.back()) + "</w>");
```

Comment:
This algorithm is essentially a 1:1 translation from the Python code and could probably be cleaned up a lot